### PR TITLE
Add Scaler

### PR DESCRIPTION
**Pull request by sheldak**
Contains Membrane element that can scale video.

### CODE REVIEW

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+  if (state->sws_context != NULL) {
+    sws_freeContext(state->sws_context);
+  }
+
+  for (int i = 0; i < 4; i++) {
+    av_freep(&state->source_data[i]);
+    av_freep(&state->scaled_data[i]);
+    av_freep(&state->target_data[i]);
+  }
```
Comment (bblaszkow06) [inline c_src/membrane_ffmpeg_swscale_plugin/scaler.c:14]: This is not the right way to free that data - `av_image_alloc` creates a continuous memory block and docs suggest using only first pointer https://ffmpeg.org/doxygen/trunk/group__lavu__picture.html#ga841e0a89a642e24141af1918a2c10448

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  uint8_t * input_data[4];
+  int input_linesize[4];
+
+  if (av_image_fill_arrays(input_data, input_linesize,
+      payload -> data, pixel_format, state -> input_width,
+      state -> input_height, 1) < 0) {
+    return scale_result_error(env, "fill_input");
+  }
+
+  sws_scale(state -> sws_context, (const uint8_t *
```
Comment (bblaszkow06) [inline c_src/membrane_ffmpeg_swscale_plugin/scaler.c:174]: Please format this file using `clang-format`. My eyes hurt seeing the spaces between struct pointer and a field :stuck_out_tongue:

File: README.md [hunk 3]
```diff
@@ -1,34 +1,56 @@
...
-  [
-    {:membrane_template_plugin, "~> 0.1.0"}
-  ]
-end
+{:membrane_ffmpeg_swscale_plugin, "~> 0.1.0"}
 ```
 
-## Usage
+You also need to have [FFmpeg](https://www.ffmpeg.org/) library installed.
+For usage on Windows, see `Using on Windows` section below.
```
Comment (bblaszkow06) [inline README.md:20]: You forgot to remove this reference along with the section

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+          dst_data[0][h * output_width + w] =
+              scaled_data[0][h * scaled_width + w - start];
+      }
+    }
+
+    for (int h = 0; h < output_height / 2; h++) {
+      for (int w = 0; w < output_width / 2; w++) {
+        if (w < start / 2 || w >= end / 2) {
+          dst_data[1][h * output_width / 2 + w] = 128;
+          dst_data[2][h * output_width / 2 + w] = 128;
```
Comment (bblaszkow06) [inline c_src/membrane_ffmpeg_swscale_plugin/scaler.c:119]: Extract this magic constant to a named one

File: lib/membrane_ffmpeg_swscale/scaler.ex [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      _state ->
+        with {:ok, frame} <- Native.scale(payload, native_state) do
+          buffer = [buffer: {:output, %{buffer | payload: frame}}]
+
+          {{:ok, buffer}, state}
+        else
+          {:error, reason} -> {{:error, reason}, state}
+        end
+    end
+  end
```
Comment (bblaszkow06) [inline lib/membrane_ffmpeg_swscale/scaler.ex:73]: Linter would complain about too much nesting. Instead of case here, do the pattern matching in a function clause

File: test/scaler/scaler_native_test.exs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  alias Membrane.FFmpeg.SWScale.Scaler.Native, as: Scaler
+
+  test "Scale 1 720p frame" do
+    in_path = "../fixtures/input-10-1280x720.raw" |> Path.expand(__DIR__)
+
+    assert {:ok, file} = File.read(in_path)
+    assert {:ok, native_state} = Scaler.create(1280, 720, 640, 640)
+    assert <<frame::bytes-size(1_382_400), _::binary>> = file
+    assert {:ok, scaled_frame} = Scaler.scale(frame, native_state)
+    assert Payload.size(scaled_frame) == 614_400
```
Comment (bblaszkow06) [inline test/scaler/scaler_native_test.exs:13]: It would be perfect to also provide a resulting file that can be compared with the output

File: test/scaler/scaling_test.exs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  defp make_pipeline_raw(in_path, out_path) do
+    Pipeline.start_link(%Pipeline.Options{
+      elements: [
+        file_src: %Membrane.File.Source{location: in_path},
+        parser: %Membrane.Element.RawVideo.Parser{format: :I420, width: 1280, height: 720},
+        scaler: %Membrane.FFmpeg.SWScale.Scaler{output_width: 400, output_height: 800},
+        sink: %Membrane.File.Sink{location: out_path}
+      ]
+    })
+  end
```
Comment (bblaszkow06) [inline test/scaler/scaling_test.exs:48]: It's a matter of opinion, but extracting this to a separate function if you're not reusing it is a bit if an overkill and forces you to make a jump while reading the test

File: test/scaler/scaler_native_test.exs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+defmodule Scaler.NativeTest do
+  use ExUnit.Case, async: true
+  alias Membrane.Payload
+  alias Membrane.FFmpeg.SWScale.Scaler.Native, as: Scaler
+
+  test "Scale 1 720p frame" do
+    in_path = "../fixtures/input-10-1280x720.raw" |> Path.expand(__DIR__)
+
+    assert {:ok, file} = File.read(in_path)
+    assert {:ok, native_state} = Scaler.create(1280, 720, 640, 640)
```
Comment (bblaszkow06) [inline test/scaler/scaler_native_test.exs:10]: This code is not fully tested - make sure to check scenarios where
- input is vertical
- input is a square
- output is vertical

in different combinations

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      }
+    }
+
+    for (int h = 0; h < output_height / 2; h++) {
+      for (int w = 0; w < output_width / 2; w++) {
+        int output_position = h * output_width / 2 + w;
+
+        if (w < start / 2 || w >= end / 2) {
+          output_data[1][output_position] = 128;
+          output_data[2][output_position] = 128;
```
Comment (bblaszkow06) [inline c_src/membrane_ffmpeg_swscale_plugin/scaler.c:124]: By "magic constant" here I meant `128`. What is it?

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+void add_paddings(uint8_t *scaled_data[4], int scaled_width, int scaled_height,
+                  uint8_t *output_data[4], int output_width,
+                  int output_height) {
+  if (scaled_height == output_height) {
+    // paddings on left and right
+    int start = (output_width - scaled_width) / 2;
+    int end = start + scaled_width;
+
+    for (int h = 0; h < output_height; h++) {
```
Comment (bblaszkow06) [inline c_src/membrane_ffmpeg_swscale_plugin/scaler.c:106]: Add a comment about this for - it's for Y plan, right? Same goes for the next one

File: README.md [hunk 11]
```diff
@@ -1,34 +1,55 @@
...
-TODO
+```elixir
+defmodule Scaling.Pipeline do
+  use Membrane.Pipeline
+
+  @doc false
+  @impl true
+  def handle_init(_) do
+    children = [
+      file_src: %Membrane.File.Source{location: "/tmp/input.raw"},
```
Comment (bblaszkow06) [inline README.md:31]: To make it clear for the user, explain above what the file should contain and what will be the result. You can also make it a bit more practical by adding decoder and encoder to the pipeline and explain that the file should contain encoded video

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: c_src/membrane_ffmpeg_swscale_plugin/scaler.c (hunk 9 above)
Comment (sheldak) [reply c_src/membrane_ffmpeg_swscale_plugin/scaler.c:124]: ahh okay. Both chrominance components need 128 to make black color. I will use `#define` for all 3 constants to make it clear

File: test/scaler/scaler_native_test.exs [hunk 12]
```diff
@@ -0,0 +1,101 @@
+defmodule Scaler.NativeTest do
+  use ExUnit.Case, async: true
+  alias Membrane.Payload
+  alias Membrane.FFmpeg.SWScale.Scaler.Native, as: Scaler
+
+  defp make_test(input_width, input_height, output_width, output_height) do
```
[the remark this answers is not in the document]
Comment (sheldak) [reply test/scaler/scaler_native_test.exs:6]: I am not sure about the convension in test files. Most of the time I see helper functions before tests like [in swresample plugin](https://github.com/membraneframework/membrane_ffmpeg_swresample_plugin/blob/master/test/membrane_ffmpeg_swresample_plugin/converter_test.exs) and [h264 plugin](https://github.com/membraneframework/membrane_h264_ffmpeg_plugin/blob/master/test/encoder/encoding_test.exs). In core (e.g. [here](https://github.com/membraneframework/membrane_core/blob/master/test/membrane/core/element_test.exs) and [here](https://github.com/membraneframework/membrane_core/blob/master/test/membrane/core/child/pad_controller_test.exs)) there are even private functions before tests. In source files I know that private functions should be at the end. But do we have some standard regarding test files, or is it arbitrary?

[the remark this answers is not in the document]
Comment (bblaszkow06) [reply test/scaler/scaler_native_test.exs:6]: Not really. In general, the functions before tests make sense if they perform a setup (often using `setup` macro from ExUnit, they cannot be private then), the helpers common for at least a couple of functions go last and helpers for a specific function (e.g. when doing recursion and having pairs like `foo` and `do_foo`) go right after the said function.
However these are only my suggestions and the way I'm used to, not a strict guideline.

One more note though - here it would make sense to keep this function on top but rename it to something like `scaling_test`

File: test/scaler/scaler_native_test.exs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  defp scale_frames(input_data, output_file, native_state, input_frame_size, scaled_frame_size) do
+    assert <<frame::bytes-size(input_frame_size), input_data::binary>> = input_data
+
+    assert {:ok, scaled_frame} = Scaler.scale(frame, native_state)
+    assert Payload.size(scaled_frame) == scaled_frame_size
+
+    assert :ok = IO.binwrite(output_file, Membrane.Payload.to_binary(scaled_frame))
+
+    scale_frames(input_data, output_file, native_state, input_frame_size, scaled_frame_size)
+  end
```
Comment (bblaszkow06) [inline test/scaler/scaler_native_test.exs:54]: You can avoid recursion using `for`
```suggestion
  defp scale_frames(input_data, output_file, native_state, input_frame_size, scaled_frame_size) do
    for <<frame::bytes-size(input_frame_size) <- input_data>> do
      assert {:ok, scaled_frame} = Scaler.scale(frame, native_state)
      assert Payload.size(scaled_frame) == scaled_frame_size

      assert :ok = IO.binwrite(output_file, Membrane.Payload.to_binary(scaled_frame))
  end

    :ok
  end
```

File: test/scaler/scaler_native_test.exs [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  use ExUnit.Case, async: true
+  alias Membrane.Payload
+  alias Membrane.FFmpeg.SWScale.Scaler.Native, as: Scaler
+
+  defp make_test(input_width, input_height, output_width, output_height) do
+    input_path = "../fixtures/input-10-#{input_width}x#{input_height}.raw" |> Path.expand(__DIR__)
+
+    output_path =
+      "../fixtures/output-10-#{input_width}x#{input_height}-to-#{output_width}x#{output_height}.raw"
+      |> Path.expand(__DIR__)
```
[the remark this answers is not in the document]
Comment (sheldak) [reply test/scaler/scaler_native_test.exs:11]: Okay, I do not need to save it to the file. I will omit it

File: README.md [hunk 15]
```diff
@@ -1,34 +1,93 @@
...
-  ]
+{:membrane_ffmpeg_swscale_plugin, "~> 0.1.0"}
+```
+
+You also need to have [FFmpeg](https://www.ffmpeg.org/) library installed.
+
+## Sample usage
+Scaler needs input in the YUV420p format. It processes one frame at a time. The element also requires
+getting caps with video width and height. To meet all requirements either `Membrane.Element.RawVideo.Parser` or `Membrane.H264.FFmpeg.Decoder`
+can be used before Scaler in the pipeline.
```
Comment (bblaszkow06) [inline README.md:24]: This seems a bit too specific - any decoder will do the job. I would rephrase it that scaler needs raw, parsed video frames so it either requires `RawVideo.Parser` or some decoder preceding it within pipeline and here you can name `H264` as an example

File: README.md [hunk 16]
```diff
@@ -1,34 +1,93 @@
...
+    ]
+
+    {{:ok, spec: %ParentSpec{children: children, links: links}}, %{}}
+  end
 end
 ```
+Result file will contain video also in YUV420p format.
+
+### H.264
+Scaling encoded (using H.264 standard) video requires parser and decoder because Scaler scales only raw video.
```
Comment (bblaszkow06) [inline README.md:56]: I'm not sure if 2 examples are needed, they are very similar, I'd only left this one.

File: lib/membrane_ffmpeg_swscale/scaler.ex [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+  @impl true
+  def handle_demand(:output, size, :buffers, _context, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, _buffer, _context, %{native_state: nil} = _state) do
+    raise(RuntimeError, :uninitialized_native_state)
```
Comment (bblaszkow06) [inline lib/membrane_ffmpeg_swscale/scaler.ex:58]: After reading the README I realized one more thing - we need to explicitly communicate what is the reason of error - like Elixir does most of the time, so I would replace this atom with a string saying the state could not be initialized because the scaler didn't receive caps. This will be a clear clue, that an error might in fact be caused by a bug in an element before or misconfigured pipeline

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: README.md [hunk 18]
```diff
@@ -1,34 +1,84 @@
...
-TODO
+## Sample usage
+
+Scaling encoded (using H.264 standard) video requires parser and decoder because Scaler scales only raw video.
+
+```elixir
+defmodule Scaling.Pipeline do
+  use Membrane.Pipeline
+
+  @doc false
```
Comment (bblaszkow06) [inline README.md:52]:
```suggestion
```
It's not needed when using `@impl true`

On a side note, be careful when writting `@impl`, `@doc` etc. in comments, you should always wrap them backticks - otherwise you'll notify someone using doc or impl as GitHub handle

**Review by bblaszkow06 — APPROVED** [review]
